provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/GradleApiExtensions.kt

```diff
@@ -0,0 +1,438 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.kotlin.dsl.codegen
+
+import java.io.File
+import java.util.Objects
+import java.util.Properties
+import java.util.jar.JarFile
+
+
+/**
+ * Generate and write Gradle API extensions to the given file.
+ *
+ * ApiTypeProvider Limitations:
+ * - supports Java byte code only, not Kotlin
+ * - does not support nested Java arrays as method parameters
+ * - does not distinguish co-variance and contra-variance
+ *
+ * GradleApiExtensions Limitations:
+ * - does not support generating extensions for functions with type parameters having multiple bounds (Kotlin where clause)
+ */
+internal
+fun writeGradleApiExtensionsTo(file: File, jars: Iterable<File>) {
+    jars.filter(gradleJarsFilter).toList().let { gradleJars ->
+        file.bufferedWriter().use {
+            it.apply {
+                write(fileHeader)
+                apiTypeProviderFor(gradleJars, parameterNamesSupplierFor(gradleJars)).use { api ->
+                    gradleApiExtensionDeclarationsFor(api).forEach {
+                        write("\n$it\n")
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+private
+val gradleJarsFilter = { jar: File ->
+    jar.name.startsWith("gradle-") && !jar.name.startsWith("gradle-kotlin-")
+}
+
+
+internal
+fun parameterNamesSupplierFor(gradleJars: List<File>) =
+    JarFile(gradleJars.single { it.name.startsWith("gradle-api-parameter-names-") }).use { jar ->
+        jar.getInputStream(jar.getJarEntry("gradle-api-parameter-names.properties")).use { input ->
+            Properties().also { it.load(input) }.let { index ->
+                { key: String ->
+                    index.getProperty(key, null)?.split(",")
+                }
+            }
+        }
+    }
+
+
+internal
+fun gradleApiExtensionDeclarationsFor(api: ApiTypeProvider): Sequence<String> =
+    api.allTypes().filter { it.isGradleApi }.flatMap { type ->
+        sequenceOf(reifiedTypeParametersExtensionsGenerator, kClassExtensionsGenerator)
+            .flatMap { generator -> generator(type) }
+            .distinctBy { it.signatureKey }
+            .map { it.toKotlinString() }
+    }
+
+
+private
+data class KotlinExtensionFunction(
+    val description: String,
+    val isIncubating: Boolean,
+    val isDeprecated: Boolean,
+    val isInline: Boolean,
+    val typeParameters: List<Pair<ApiTypeUsage, Boolean>>,
+    val targetType: ApiType,
+    val name: String,
+    val parameters: List<ApiFunctionParameter>,
+    val returnType: ApiTypeUsage,
+    val expressionBody: String
+) {
+
+    val signatureKey: Int = Objects.hash(targetType.sourceName, name, parameters.map { it.type })
+
+    fun toKotlinString(): String = StringBuilder().apply {
+        appendln("""
+            /**
+             * $description.
```

Comment:
Can the `description` contain a `*/` within it?

Reply:
There're only two private call sites below, I don't see a need for escaping